### ISSUE TITLE
Failure to parse "value as Type[index]"

### ISSUE THREAD

**Issue by CensoredUsername** [rust-lang/rust#35813]
Summary:

Parsing `value as Type[index]` leads to the error:

```
expected one of `!`, `(`, `::`, `;`, or `<`, found `[`
```

It is currently impossible for a cast expression to have a trailing index expression without surrounding the cast expression by parenthesis. This however seems to be a parser limitation, not a grammar limitation.

As currently the macro language guarantees that ty/path expressions can be followed by a "[", this seems like a bug in the parser to me, as it seems to be parsing it as `value as (Type[index])` which makes no grammatical sense, compared to parsing it as (value as Type)[index] which should be valid if I read the lexical structure specification correctly.

```
rustc --version
>>> rustc 1.12.0-nightly (576f76659 2016-08-09)
```

**Comment by oli-obk** [comment]
I think it's parsed correctly, but should get a tailored error message to suggest surrounding the cast with parens.

**Comment by CensoredUsername** [comment]
Considering that a trailing "[" is never valid in the type syntax (otherwise macros break) and the fact that `foo as Type * bar` and variants are perfectly legal (even `>` which is another symbol that occurs in type syntax).

But if a tailored error message will be made instead, then this should also include other postfix operators like `val as Type . attr` which are currently rejected even though technically being legal by the lexical grammar specification. (another one is the proposed "?" operator)

**Comment by Mark-Simulacrum** [comment]
This is still lacking a tailored error message, and as far as I know no movement has been made here.

**Comment by daboross** [comment]
Submitted a pull request to fix this! #68985

**Comment by CensoredUsername** [comment]
Thanks!
